sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -62,6 +62,7 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
       case _ =>
         Set.empty[Expression]
     }.foldLeft(Set.empty[Expression])(_ union _.toSet)
+      .union(output.filterNot(_.nullable).map(IsNotNull).toSet)
```

Comment:
can we break chains like this down with intermediate variables and comments? they are becoming fairly difficult to understand.

Also the foldLeft/union ... seems like a pretty inefficient way to run this.